File: 2807. Insert Greatest Common Divisors in Linked List/2807. Insert Greatest Common Divisors in Linked List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var insertGreatestCommonDivisors = function (head) {
  const gcd = (a, b) => {
    while (b) {
      [a, b] = [b, a % b];
    }
    return a;
  };

  curr = head;
  while (curr && curr.next) {
    const gcdVal = gcd(curr.val, curr.next.val);
    curr.next = new ListNode(gcdVal, curr.next);
    curr = curr.next.next;
  }
  return head;
};

File: 2807. Insert Greatest Common Divisors in Linked List/2807. Insert Greatest Common Divisors in Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {

  public ListNode insertGreatestCommonDivisors(ListNode head) {
    ListNode curr = head;
    while (curr != null && curr.next != null) {
      int gcdVal = gcd(curr.val, curr.next.val);
      ListNode temp = curr.next;
      curr.next = new ListNode(gcdVal, curr.next);
      curr.next.next = temp;
      curr = curr.next.next;
    }
    return head;
  }

  private int gcd(int a, int b) {
    while (b > 0) {
      int temp = a;
      a = b;
      b = temp % b;
    }
    return a;
  }
}
